### PR TITLE
[WIP] Async filters feature branch

### PR DESCRIPTION
**Pull request by pacrob**
### What was wrong?
Filters need async, this branch will collect related changes.

Related to Issue #2573

### Todo:
- [ ] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

File: web3/_utils/module_testing/eth_module.py [hunk 1]
```diff
@@ -1405,6 +1405,74 @@ def test_async_provider_default_block(
...
+        filter = await async_w3.eth.filter("latest")
+        assert is_string(filter.filter_id)
+
+        changes = await async_w3.eth.get_filter_changes(
+            filter.filter_id
+        )  # type: ignore
+        assert is_list_like(changes)
+        assert not changes
+
+        # TODO: figure out why this fails in go-ethereum
```
Comment (fselmo) [inline web3/_utils/module_testing/eth_module.py:1436]: We can take these TODOs out. Looks like it fails right [here](https://github.com/ethereum/go-ethereum/blob/b196ad1c165ecd6c9edaca520e7161a58e50eb06/eth/filters/api.go#L367). If I'm reading this correctly, neither a `BlocksSubscription` (created via `w3.eth.filter("latest")`) nor a `PendingTransactionsSubscription` (created via `w3.eth.filter("pending")`) is accepted for `eth_getLogs`. Only a `LogsSubscription` (or `eth_newFilter`... created by supplying the `filter_params` dict) is supported and we have tests for that below.

We can take these out from the sync tests as well.

**Review by fselmo — COMMENTED** [review]
I left a comment on the TODOs. We can remove those. I'd like to add testing around the logs as well. Did you run into any issues porting those from sync to async?

**Review by fselmo — DISMISSED** [review]
> I left a comment on the TODOs. We can remove those. I'd like to add testing around the logs as well. Did you run into any issues porting those from sync to async?

Disregard that, those were `eth_getLogs` tests I was seeing, not `eth_getFilterLogs`. Taking out these TODOs, This looks good to me 👍

**Comment by fselmo** [comment]
Also.... don't forget to add a newsfragment and squashing both commits before merging 👀

**Comment by fselmo** [comment]
Oops... I guess if this is the base for the feature branch I will un-approve so we can approve different PRs into here and approve the whole thing later.

File: web3/_utils/module_testing/eth_module.py (hunk 1 above)
Comment (pacrob) [reply web3/_utils/module_testing/eth_module.py]: Removed!

**Pull request merged**